hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationStateHolder.java

```diff
@@ -99,10 +116,52 @@ void prepare(PartitionContainer container, Collection<ServiceNamespace> namespac
                 recordSet.add(recordReplicationInfo);
             }
             data.put(mapName, recordSet);
+
+            Set<IndexInfo> indexInfos = new HashSet<IndexInfo>();
+            if (mapContainer.isGlobalIndexEnabled()) {
+                // global-index
+                for (Index index : mapContainer.getIndexes().getIndexes()) {
+                    indexInfos.add(new IndexInfo(index.getAttributeName(), index.isOrdered()));
+                }
+            } else {
+                // partitioned-index
+                for (PartitionContainer partitionContainer : mapServiceContext.getPartitionContainers()) {
+                    final Indexes indexes = mapContainer.getIndexes(partitionContainer.getPartitionId());
+                    if (indexes != null && indexes.hasIndex()) {
+                        for (Index index : indexes.getIndexes()) {
+                            indexInfos.add(new IndexInfo(index.getAttributeName(), index.isOrdered()));
+                        }
+                    }
+                }
+            }
+            MapIndexInfo mapIndexInfo = new MapIndexInfo(mapName);
+            mapIndexInfo.addIndexInfos(indexInfos);
+            mapIndexInfos.add(mapIndexInfo);
         }
+
     }
 
     void applyState() {
+        assert Thread.currentThread() instanceof PartitionOperationThread;
+
+        // the null check can be removed in 3.10+ codebase
+        if (mapIndexInfos != null) {
+            for (MapIndexInfo mapIndexInfo : mapIndexInfos) {
+                addIndexes(mapIndexInfo.getMapName(), mapIndexInfo.getLsIndexes());
+            }
+        }
+
+        // RU_COMPAT_38
+        // Old nodes (3.8-) won't send mapIndexInfos to new nodes (3.9+) in the map-replication operation.
+        // This is the reason why we pick up the mapContainer.getIndexesToAdd() that were added by the PostJoinMapOperation
+        // and we add them to the map, before we add data
+        for (String mapName : data.keySet()) {
+            RecordStore recordStore = mapReplicationOperation.getRecordStore(mapName);
+            MapContainer mapContainer = recordStore.getMapContainer();
+            addIndexes(mapName, mapContainer.getIndexesToAdd());
+        }
+
+
         if (data != null) {
```

Comment:
Do we still need to add the indexes again on line 174?
```
for (Map.Entry<String, Boolean> indexDefinition : mapContainer.getIndexDefinitions().entrySet()) {
                    Indexes indexes = mapContainer.getIndexes(partitionContainer.getPartitionId());
                    indexes.addOrGetIndex(indexDefinition.getKey(), indexDefinition.getValue());
                }
```

Reply:
Mind that we are adding from two different locations. 
Line 159 is the compat thingie, since we won't get the indexes in the line 72.
Line 174 is used when index are sent in line 72.